packages/network-controller/tests/fake-provider-engine.ts

```diff
@@ -0,0 +1,208 @@
+ * @property response.error - Specifies an error response, with this as the
+ * `error`.
+ * @property error - Instructs `sendAsync` to return a promise that rejects with
+ * this error.
+ * @property implementation - Allows overriding `sendAsync` entirely. Useful if
+ * you want it to throw an error.
+ * @property delay - The amount of time that will pass after the callback is
+ * called with the response.
+ * @property discardAfterMatching - Usually after the stub matches a request, it
+ * is discarded, but setting this to true prevents that from happening. True by
+ * default.
+ */
+export type FakeProviderStub = {
+  request: {
+    method: string;
+    params?: (string | boolean)[];
+  };
+  delay?: number;
+  discardAfterMatching?: boolean;
+} & (
+  | {
+      response: { result: any } | { error: string };
+    }
+  | {
+      error: string;
+    }
+  | {
+      implementation: () => void;
+    }
+);
+
+/**
+ * The set of options that the FakeProviderEngine constructor takes.
+ *
+ * @property stubs - A set of objects that allow specifying the behavior
+ * of specific invocations of `sendAsync` matching a `method`.
+ */
+interface FakeProviderEngineOptions {
+  stubs?: FakeProviderStub[];
+}
+
+/**
+ * FakeProviderEngine is an implementation of the provider that
+ * NetworkController exposes, which is actually an instance of
+ * Web3ProviderEngine (from the `web3-provider-engine` package). Hence it
+ * supports the same interface as Web3ProviderEngine, except that fake responses
+ * for any RPC methods that are accessed can be supplied via an API that is more
+ * succinct than using Jest's mocking API.
+ */
+export class FakeProviderEngine extends EventEmitter implements ProviderEngine {
+  #isStopped: boolean;
+
+  #stubs: FakeProviderStub[];
+
+  calledStubs: FakeProviderStub[];
+
+  #originalStubs: FakeProviderStub[];
+
+  /**
+   * Makes a new instance of the fake provider.
+   *
+   * @param options - The options.
+   * @param options.stubs - A set of objects that allow specifying the behavior
+   * of specific invocations of `sendAsync` matching a `method`.
+   */
+  constructor({ stubs = [] }: FakeProviderEngineOptions) {
+    super();
+    this.#originalStubs = stubs;
+    this.#stubs = this.#originalStubs.slice();
+    this.calledStubs = [];
+    this.#isStopped = false;
+  }
+
+  stop() {
+    this.#isStopped = true;
+  }
+
+  sendAsync<P, V>(
+    payload: SendAsyncPayload<P>,
+    callback: (error: unknown, response: RpcResponse<RpcPayload<P>, V>) => void,
+  ) {
+    if (Array.isArray(payload)) {
+      throw new Error("Arrays aren't supported");
+    }
+
+    const index = this.#stubs.findIndex((stub) => {
+      return (
+        stub.request.method === payload.method &&
+        (!('params' in stub.request) ||
+          isDeepStrictEqual(stub.request.params, payload.params))
+      );
+    });
+
+    if (this.#isStopped) {
+      console.error(`The provider has been stopped, yet sendAsync has somehow been called. If this
```

Comment:
I'm seeing this printed to the console repeatedly when running tests. Is this anticipated?

Reply:
Yes, and I see it being printed repeatedly on my side as well. I believe this is happening because of how `lookupNetwork` and `getEIP1559Compatibility` are used. I feel like it is useful to know that calls to `sendAsync` are happening even when the provider is stopped, but not sure how others may see this. Do you find it helpful or should I remove this?

Reply:
Agreed that it is useful, but the use of `error` level implies that this is not supposed to be happening.

I will review this a bit further to see if I can understand the cause, but if we can't solve it here maybe `warn` is more appropriate.

Reply:
I looked into this more, and just as on the extension, it's caused by not awaiting for all operations triggered by the code under test. For example for the `set providerConfig` tests, we'd want to wait until the network lookup was finished, and the EIP compatibility check, and so on.

This is fixable in this PR, but I think it would be more pragmatic to leave it as-is for now and fix this in a later PR after we clean up the API and make it easy to `await` each operation. In the meantime, what do you think of making this `console.warn` instead?

Reply:
Yup! Updated to use `console.warn` here: [9f27bf6](https://github.com/MetaMask/core/pull/1039/commits/9f27bf6f6d4f859e43e85efafb8de36309d74080)